app/services/journal_row_builder.rb

```diff
@@ -0,0 +1,153 @@
+require "set"
+
+class JournalRowBuilder
+
+  attr_reader :order_details, :journal, :journal_rows, :errors,
+    :recharge_enabled, :product_recharges, :journaled_facility_ids,
+    :pending_facility_ids
+
+  # Minimal initialization.
+  def initialize(order_details, journal)
+    @journal = journal
+    @order_details = order_details
+    @recharge_enabled = SettingsHelper.feature_on?(:recharge_accounts)
+    reset
+  end
+
+  # Idempotent method that builds an array of new journal_rows. Also adds to
+  # errors if any problems occur. Does not save anything to the database.
+  # Returns self.
+  def build
+    reset
+    cache_pending_facility_ids
+    order_details.each do |order_detail|
+      account = order_detail.account
+      validate_order_detail_unjournaled(order_detail)
+      validate_facility_can_journal(order_detail)
+      validate_account(order_detail)
+      new_journal_rows = order_detail_to_journal_rows(order_detail)
+      update_product_recharges(order_detail)
+      @journal_rows.concat(new_journal_rows)
+    end
+    add_journal_rows_from_product_recharges
+    self
+  end
+
+  private
+
+  # Reset instance variables both on initialize and build.
+  def reset
+    @errors = []
+    @journal_rows = []
+    @journaled_facility_ids = Set.new
+    @pending_facility_ids = []
+    @product_recharges = {}
+  end
+
+  # Set an array of facility_ids with pending journals
+  def cache_pending_facility_ids
```

Comment:
Instead of a `cache` method and the attr_reader, I would memoize this
```ruby
def pending_facility_ids
  @pending_facility_ids ||= Journal.facility_ids_with_pending_journals
end
```

Also, how this is used seems kind of weird (it was weird previously as well). Could we do this check ahead of time? Error out if any order_details have a facility with a pending journal: `(order_details.distinct.pluck(:facility_id) - Journal.facility_ids_with_pending_journals).any?` (or equivalent--I think we don't get `distinct.pluck` in 3.2).